hledger-lib/Hledger/Data/Dates.hs

```diff
@@ -741,25 +865,58 @@ reportinginterval = choice' [
                           many spacenonewline
                           string "day"
                           many spacenonewline
-                          string "of"
+                          of_ "week"
+                          return $ DayOfWeek n,
+                       do string "every"
+                          many spacenonewline
+                          n <- weekday
                           many spacenonewline
-                          string "week"
                           return $ DayOfWeek n,
                        do string "every"
                           many spacenonewline
                           n <- fmap read $ some digitChar
                           thsuffix
                           many spacenonewline
                           string "day"
-                          optional $ do
-                            many spacenonewline
-                            string "of"
-                            many spacenonewline
-                            string "month"
-                          return $ DayOfMonth n
+                          many spacenonewline
+                          optional $ of_ "month"
+                          return $ DayOfMonth n,
+                       do string "every"
+                          many spacenonewline
+                          d_o_y <- makePermParser $
+                               DayOfYear <$$> do (_,m,_) <- choice' [month, mon]
+                                                 many spacenonewline
+                                                 return $ read m
+                                                 
+                                         <||> do d <- fmap read $ some digitChar
+                                                 thsuffix
+                                                 many spacenonewline
+                                                 return d
+                          optional $ of_ "year"
+                          return d_o_y,
+                       do string "every"
+                          many spacenonewline
+                          ("",m,d) <- md
+                          many spacenonewline
+                          optional $ of_ "year"
+                          return $ DayOfYear (read m) (read d),
+                       do string "every"
+                          many spacenonewline
+                          n <- fmap read $ some digitChar
```

Comment:
Since you've introduced `of_` you can add:

```haskell
nth = read <$> some digitChar <* thsuffix
```

Reply:
Sensible. Done